File: desktop/buddy_ai/voice.py
"""Conversation state machine.

Listening uses the real microphone level and replies are spoken with the
operating system's text-to-speech (see ``audio.py``). There is no speech
recognition: the reply is a generic one chosen by how much was said. When
live audio is unavailable, ``MicSimulator`` / ``SpeechSimulator`` fake it.
"""

from __future__ import annotations

import math
import random
import re
import time
from concurrent.futures import Future, ThreadPoolExecutor
from dataclasses import dataclass
from enum import Enum, auto
from typing import Callable

from . import audio


class Mode(Enum):
    IDLE = auto()
    LISTENING = auto()
    THINKING = auto()
    SPEAKING = auto()


NO_SPEECH_REPLIES = (
    "Hmm, I didn't quite catch that, {name}. Could you say it again?",
    "I'm here and listening, {name}. Whenever you're ready, just speak up.",
)
SHORT_REPLIES = (
    "Got it, {name}! Thanks for letting me know.",
    "Okay, understood. Is there anything else on your mind?",
    "Nice! I hear you loud and clear.",
)
LONG_REPLIES = (
    "Thanks for sharing all of that, {name}. That sounds really thoughtful. Tell me more whenever you like.",
    "I hear you. That's a lot to think about, and you explained it really well.",
    "Great point, {name}. I love where you're going with this. Let's keep talking.",
)


def approach(current: float, target: float, rate: float, dt: float) -> float:
    """Frame-rate independent exponential smoothing."""
    return current + (target - current) * (1.0 - math.exp(-rate * dt))


def count_syllables(word: str) -> int:
    word = word.lower()
    groups = re.findall(r"[aeiouy]+", word)
    silent_e = word.rstrip(".,!?;:'\"").endswith("e") and len(groups) > 1
    return max(1, len(groups) - (1 if silent_e else 0))


class MicSimulator:
    """Fakes a microphone level: bursts of talking separated by pauses."""

    def __init__(self, rng: random.Random) -> None:
        self._rng = rng
        self._talking = False
        self._segment_left = rng.uniform(0.3, 0.6)
        self._retarget_in = 0.0
        self._target = 0.0
        self.level = 0.0

    def step(self, dt: float) -> float:
        rng = self._rng
        self._segment_left -= dt
        if self._segment_left <= 0:
            self._talking = not self._talking
            self._segment_left = rng.uniform(0.7, 2.2) if self._talking else rng.uniform(0.25, 0.9)
        self._retarget_in -= dt
        if self._retarget_in <= 0:
            self._retarget_in = rng.uniform(0.05, 0.15)
            self._target = rng.uniform(0.35, 1.0) if self._talking else rng.uniform(0.0, 0.05)
        self.level = approach(self.level, self._target, 20.0, dt)
        return self.level


@dataclass
class TimedWord:
    text: str
    start: float
    end: float
    syllables: int
    peak: float


class SpeechSimulator:
    """Schedules a reply word by word and produces a syllable-shaped level."""

    def __init__(self, text: str, rng: random.Random) -> None:
        self.text = text
        self.words: list[TimedWord] = []
        self._rng = rng
        t = 0.3
        for raw in text.split():
            syllables = count_syllables(raw)
            duration = 0.10 + 0.16 * syllables
            self.words.append(TimedWord(raw, t, t + duration, syllables, rng.uniform(0.6, 1.0)))
            t += duration + 0.05
            if raw[-1] in ",;:":
                t += 0.22
            elif raw[-1] in ".!?":
                t += 0.45
        self.duration = t + 0.2
        self.elapsed = 0.0

    @property
    def done(self) -> bool:
        return self.elapsed >= self.duration

    def step(self, dt: float) -> float:
        self.elapsed += dt
        for word in self.words:
            if word.start <= self.elapsed < word.end:
                progress = (self.elapsed - word.start) / (word.end - word.start)
                phase = (progress * word.syllables) % 1.0
                return word.peak * math.sin(math.pi * phase) ** 0.6 * self._rng.uniform(0.85, 1.0)
        return 0.0

    def word_progress(self) -> list[float]:
        """How far each word has been spoken, in [0, 1]."""
        out = []
        for word in self.words:
            if self.elapsed <= word.start:
                out.append(0.0)
            elif self.elapsed >= word.end:
                out.append(1.0)
            else:
                out.append((self.elapsed - word.start) / (word.end - word.start))
        return out


class SpokenReply(SpeechSimulator):
    """A reply backed by synthesized audio; captions are stretched to fit it."""

    def __init__(self, text: str, rng: random.Random, utterance: audio.Utterance) -> None:
        super().__init__(text, rng)
        self.utterance = utterance
        stretch = utterance.duration / max(self.duration, 1e-6)
        for word in self.words:
            word.start *= stretch
            word.end *= stretch
        self.duration = utterance.duration + 0.25
        self._started_at: float | None = None

    def step(self, dt: float) -> float:
        if self._started_at is None:
            audio.play(self.utterance)
            self._started_at = time.monotonic()
        self.elapsed = time.monotonic() - self._started_at
        return self.utterance.level_at(self.elapsed)

    def stop(self) -> None:
        audio.stop_playback()


class Conversation:
    """IDLE -> LISTENING -> THINKING -> SPEAKING -> IDLE, driven by ``step``."""

    MIN_THINK_SECONDS = (0.7, 1.2)
    MAX_THINK_SECONDS = 12.0
    SPEECH_LEVEL = 0.28      # mic level that counts as talking
    END_OF_TURN_SILENCE = 1.5
    NO_SPEECH_TIMEOUT = 12.0

    def __init__(self, name: str = "friend", seed: int | None = None, live_audio: bool = True) -> None:
        self.name = name
        self.mode = Mode.IDLE
        self.mode_time = 0.0
        self.level = 0.0
        self.speech: SpeechSimulator | None = None
        self.voice_hint = ""
        self.voice_rate = 1
        self._rng = random.Random(seed)
        self._live_mic = audio.LiveMic() if live_audio and audio.available() else None
        self._speak_aloud = live_audio and audio.available()
        self._using_live_mic = False
        self._mic: MicSimulator | None = None
        self._talk_time = 0.0
        self._silence = 0.0
        self._think_left = 0.0
        self._reply_text = ""
        self._pending: Future | None = None
        self._executor = ThreadPoolExecutor(max_workers=1)
        self._last_reply: str | None = None
        self._listeners: list[Callable[[Mode], None]] = []
        self._notice_listeners: list[Callable[[str], None]] = []

    @property
    def using_live_mic(self) -> bool:
        return self._using_live_mic

    def on_mode_changed(self, callback: Callable[[Mode], None]) -> None:
        self._listeners.append(callback)

    def on_notice(self, callback: Callable[[str], None]) -> None:
        self._notice_listeners.append(callback)

    def toggle_mic(self) -> None:
        """Mic tap: start listening, or finish the turn if already listening."""
        self._set_mode(Mode.THINKING if self.mode is Mode.LISTENING else Mode.LISTENING)

    def cancel(self) -> None:
        self._set_mode(Mode.IDLE)

    def shutdown(self) -> None:
        self.cancel()
        self._executor.shutdown(wait=False, cancel_futures=True)

    def step(self, dt: float) -> None:
        self.mode_time += dt
        raw = 0.0
        if self.mode is Mode.LISTENING:
            raw = self._live_mic.level if self._using_live_mic else self._mic.step(dt)
            self._track_turn(raw, dt)
        elif self.mode is Mode.THINKING:
            raw = 0.12 + 0.06 * math.sin(self.mode_time * 6.0)
            self._think_left -= dt
            ready = self._pending is None or self._pending.done() or self.mode_time > self.MAX_THINK_SECONDS
            if self._think_left <= 0 and ready:
                self._set_mode(Mode.SPEAKING)
        elif self.mode is Mode.SPEAKING and self.speech:
            raw = self.speech.step(dt)
            if self.speech.done:
                self._set_mode(Mode.IDLE)
        self.level = approach(self.level, raw, 14.0, dt)

    def _track_turn(self, raw: float, dt: float) -> None:
        if raw > self.SPEECH_LEVEL:
            self._talk_time += dt
            self._silence = 0.0
        else:
            self._silence += dt
        if not self._using_live_mic:
            return
        finished_talking = self._talk_time > 0.5 and self._silence > self.END_OF_TURN_SILENCE
        gave_up = self._talk_time < 0.3 and self.mode_time > self.NO_SPEECH_TIMEOUT
        if finished_talking or gave_up:
            self._set_mode(Mode.THINKING)

    def _set_mode(self, mode: Mode) -> None:
        if mode is self.mode:
            return
        self._leave(self.mode)
        self.mode = mode
        self.mode_time = 0.0
        if mode is Mode.LISTENING:
            self._start_listening()
        elif mode is Mode.THINKING:
            self._start_thinking()
        elif mode is Mode.SPEAKING:
            self._start_speaking()
        for callback in self._listeners:
            callback(mode)

    def _leave(self, mode: Mode) -> None:
        if mode is Mode.LISTENING and self._using_live_mic:
            self._live_mic.stop()
            self._using_live_mic = False
        elif mode is Mode.SPEAKING and isinstance(self.speech, SpokenReply):
            self.speech.stop()

    def _start_listening(self) -> None:
        self.speech = None
        self._pending = None
        self._talk_time = self._silence = 0.0
        self._mic = MicSimulator(self._rng)
        if self._live_mic is not None:
            try:
                self._live_mic.start()
                self._using_live_mic = True
            except Exception:
                self._live_mic = None
                self._notice("Microphone unavailable, simulating your voice")

    def _start_thinking(self) -> None:
        self._think_left = self._rng.uniform(*self.MIN_THINK_SECONDS)
        self._reply_text = self._pick_reply()
        self._pending = None
        if self._speak_aloud:
            self._pending = self._executor.submit(
                audio.synthesize, self._reply_text, self.voice_hint, self.voice_rate)

    def _start_speaking(self) -> None:
        utterance, timed_out = None, False
        if self._pending is not None:
            if self._pending.done() and not self._pending.cancelled():
                utterance = self._pending.result()
            else:
                timed_out = True
        self._pending = None
        if utterance is not None:
            self.speech = SpokenReply(self._reply_text, self._rng, utterance)
            return
        if self._speak_aloud and not timed_out:
            self._speak_aloud = False
            self._notice("Text-to-speech unavailable, replying with captions only")
        self.speech = SpeechSimulator(self._reply_text, self._rng)

    def _pick_reply(self) -> str:
        if self._talk_time < 0.3:
            pool = NO_SPEECH_REPLIES
        elif self._talk_time < 2.0:
            pool = SHORT_REPLIES
        else:
            pool = LONG_REPLIES
        options = [r for r in pool if r != self._last_reply]
        self._last_reply = self._rng.choice(options)
        return self._last_reply.format(name=self.name)

    def _notice(self, message: str) -> None:
        for callback in self._notice_listeners:
            callback(message)

File: src/buddy/voice.ts
/**
 * Conversation state machine: idle -> listening -> thinking -> speaking -> idle.
 *
 * Audio comes in through small interfaces (mic, speaker, recognizer) so the
 * browser implementations in `audio.ts` can be swapped for fakes in tests.
 * Anything that is missing or fails is replaced by a simulation, so the
 * buddy always has something believable to react to.
 */

export type Mode = 'idle' | 'listening' | 'thinking' | 'speaking'
export const MODES: readonly Mode[] = ['idle', 'listening', 'thinking', 'speaking']

export type Rng = () => number

/** Small deterministic PRNG (mulberry32) for reproducible tests. */
export function seededRng(seed: number): Rng {
  let a = seed >>> 0
  return () => {
    a = (a + 0x6d2b79f5) >>> 0
    let t = a
    t = Math.imul(t ^ (t >>> 15), t | 1)
    t ^= t + Math.imul(t ^ (t >>> 7), t | 61)
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296
  }
}

export const uniform = (rng: Rng, lo: number, hi: number) => lo + (hi - lo) * rng()

/** Frame-rate independent exponential smoothing. */
export function approach(current: number, target: number, rate: number, dt: number): number {
  return current + (target - current) * (1 - Math.exp(-rate * dt))
}

export function countSyllables(word: string): number {
  const w = word.toLowerCase()
  const groups = w.match(/[aeiouy]+/g) ?? []
  const silentE = w.replace(/[.,!?;:'"]+$/, '').endsWith('e') && groups.length > 1
  return Math.max(1, groups.length - (silentE ? 1 : 0))
}

export const NO_SPEECH_REPLIES = [
  "Hmm, I didn't quite catch that, {name}. Could you say it again?",
  "I'm here and listening, {name}. Whenever you're ready, just speak up.",
]
export const SHORT_REPLIES = [
  'Got it, {name}! Thanks for letting me know.',
  'Okay, understood. Is there anything else on your mind?',
  'Nice! I hear you loud and clear.',
]
export const LONG_REPLIES = [
  'Thanks for sharing all of that, {name}. That sounds really thoughtful. Tell me more whenever you like.',
  "I hear you. That's a lot to think about, and you explained it really well.",
  "Great point, {name}. I love where you're going with this. Let's keep talking.",
]
/** Used when speech recognition caught the words, so the buddy can quote them. */
export const HEARD_REPLIES = [
  'I heard you say, "{snippet}". Thanks for sharing that, {name}!',
  '"{snippet}". Got it, {name}. Tell me more whenever you like.',
]

export interface ReplyContext {
  name: string
  talkTime: number
  transcript: string
  last?: string
}

/** Pick a generic reply based on how much was said (and what, when known). */
export function chooseReply(rng: Rng, ctx: ReplyContext): { template: string; text: string } {
  const words = ctx.transcript.split(/\s+/).filter(Boolean)
  let pool: string[]
  if (words.length >= 3 && rng() < 0.6) pool = HEARD_REPLIES
  else if (words.length === 0 && ctx.talkTime < 0.3) pool = NO_SPEECH_REPLIES
  else if (words.length > 0 ? words.length <= 4 : ctx.talkTime < 2) pool = SHORT_REPLIES
  else pool = LONG_REPLIES

  const options = pool.length > 1 ? pool.filter((r) => r !== ctx.last) : pool
  const template = options[Math.floor(rng() * options.length)] ?? pool[0]!
  const snippet =
    words
      .slice(0, 6)
      .join(' ')
      .replace(/[.,!?;:]+$/, '') + (words.length > 6 ? '...' : '')
  const text = template.replaceAll('{name}', ctx.name).replaceAll('{snippet}', snippet)
  return { template, text }
}

/** Fakes a microphone level: bursts of talking separated by pauses. */
export class MicSimulator {
  level = 0
  private talking = false
  private segmentLeft: number
  private retargetIn = 0
  private target = 0

  constructor(private readonly rng: Rng) {
    this.segmentLeft = uniform(rng, 0.3, 0.6)
  }

  step(dt: number): number {
    const rng = this.rng
    this.segmentLeft -= dt
    if (this.segmentLeft <= 0) {
      this.talking = !this.talking
      this.segmentLeft = this.talking ? uniform(rng, 0.7, 2.2) : uniform(rng, 0.25, 0.9)
    }
    this.retargetIn -= dt
    if (this.retargetIn <= 0) {
      this.retargetIn = uniform(rng, 0.05, 0.15)
      this.target = this.talking ? uniform(rng, 0.35, 1) : uniform(rng, 0, 0.05)
    }
    this.level = approach(this.level, this.target, 20, dt)
    return this.level
  }
}

export interface TimedWord {
  text: string
  charIndex: number
  start: number
  end: number
  syllables: number
  peak: number
}

/**
 * Word-by-word schedule for a reply. Drives the captions and a
 * syllable-shaped mouth level; resynced from the voice's word boundaries.
 */
export class SpeechTimeline {
  readonly words: TimedWord[] = []
  readonly duration: number
  elapsed = 0

  constructor(
    readonly text: string,
    private readonly rng: Rng,
    rate = 1,
  ) {
    let t = 0.3
    for (const match of text.matchAll(/\S+/g)) {
      const raw = match[0]
      const syllables = countSyllables(raw)
      const duration = (0.1 + 0.16 * syllables) / rate
      this.words.push({
        text: raw,
        charIndex: match.index,
        start: t,
        end: t + duration,
        syllables,
        peak: uniform(rng, 0.6, 1),
      })
      t += duration + 0.05 / rate
      const last = raw.at(-1) ?? ''
      if (',;:'.includes(last)) t += 0.22 / rate
      else if ('.!?'.includes(last)) t += 0.45 / rate
    }
    this.duration = t + 0.2
  }

  get done(): boolean {
    return this.elapsed >= this.duration
  }

  step(dt: number): number {
    this.elapsed += dt
    for (const w of this.words) {
      if (w.start <= this.elapsed && this.elapsed < w.end) {
        const phase = (((this.elapsed - w.start) / (w.end - w.start)) * w.syllables) % 1
        return w.peak * Math.sin(Math.PI * phase) ** 0.6 * uniform(this.rng, 0.85, 1)
      }
    }
    return 0
  }

  /** The voice just started the word at `charIndex`: snap the schedule to it. */
  syncToChar(charIndex: number): void {
    const word = this.words.findLast((w) => w.charIndex <= charIndex)
    if (word && Math.abs(this.elapsed - word.start) > 0.12) this.elapsed = word.start
  }

  finish(): void {
    this.elapsed = Math.max(this.elapsed, this.duration)
  }

  wordProgress(): number[] {
    return this.words.map((w) =>
      this.elapsed <= w.start
        ? 0
        : this.elapsed >= w.end
          ? 1
          : (this.elapsed - w.start) / (w.end - w.start),
    )
  }
}

export interface MicInput {
  start(): Promise<void>
  stop(): void
  /** Current loudness in [0, 1]; call once per frame. */
  readLevel(): number
}

export interface VoiceStyle {
  voiceIndex: number
  rate: number
  pitch: number
}

export interface SpeechCallbacks {
  onBoundary(charIndex: number): void
  onEnd(): void
  onError(): void
}

export interface SpeechOutput {
  speak(text: string, style: VoiceStyle, callbacks: SpeechCallbacks): void
  stop(): void
}

export interface Recognizer {
  start(onText: (text: string) => void): void
  stop(): void
}

export interface ConversationOptions {
  name?: string
  rng?: Rng
  mic?: MicInput | null
  speaker?: SpeechOutput | null
  recognizer?: Recognizer | null
}

type TtsState = 'none' | 'playing' | 'ended' | 'failed'

export class Conversation {
  static readonly THINK_SECONDS: [number, number] = [0.9, 1.6]
  static readonly SPEECH_LEVEL = 0.28 // mic level that counts as talking
  static readonly END_OF_TURN_SILENCE = 1.5
  static readonly NO_SPEECH_TIMEOUT = 12
  static readonly VOICE_GRACE = 6 // extra seconds to wait for a slow voice

  mode: Mode = 'idle'
  modeTime = 0
  level = 0
  transcript = ''
  speech: SpeechTimeline | null = null
  voice: VoiceStyle = { voiceIndex: 0, rate: 1, pitch: 1 }
  name: string

  private readonly rng: Rng
  private mic: MicInput | null
  private speaker: SpeechOutput | null
  private readonly recognizer: Recognizer | null
  private usingLiveMic = false
  private liveTime = 0
  private sim: MicSimulator | null = null
  private talkTime = 0
  private silence = 0
  private thinkLeft = 0
  private replyText = ''
  private lastTemplate: string | undefined
  private tts: TtsState = 'none'
  private session = 0
  private readonly modeListeners: ((mode: Mode) => void)[] = []
  private readonly noticeListeners: ((message: string) => void)[] = []

  constructor(options: ConversationOptions = {}) {
    this.name = options.name ?? 'friend'
    this.rng = options.rng ?? Math.random
    this.mic = options.mic ?? null
    this.speaker = options.speaker ?? null
    this.recognizer = options.recognizer ?? null
  }

  get liveMic(): boolean {
    return this.usingLiveMic
  }

  onModeChanged(callback: (mode: Mode) => void): void {
    this.modeListeners.push(callback)
  }

  onNotice(callback: (message: string) => void): void {
    this.noticeListeners.push(callback)
  }

  /** Mic tap: start listening, or finish the turn if already listening. */
  toggleMic(): void {
    this.setMode(this.mode === 'listening' ? 'thinking' : 'listening')
  }

  cancel(): void {
    this.setMode('idle')
  }

  step(dt: number): void {
    this.modeTime += dt
    let raw = 0
    if (this.mode === 'listening') {
      if (this.usingLiveMic && this.mic) raw = this.mic.readLevel()
      else if (!this.mic && this.sim) raw = this.sim.step(dt) // no mic at all: simulate
      this.trackTurn(raw, dt)
    } else if (this.mode === 'thinking') {
      raw = 0.12 + 0.06 * Math.sin(this.modeTime * 6)
      this.thinkLeft -= dt
      if (this.thinkLeft <= 0) this.setMode('speaking')
    } else if (this.mode === 'speaking' && this.speech) {
      const s = this.speech
      if (this.tts === 'ended') {
        s.finish()
        this.setMode('idle')
      } else {
        raw = s.step(dt)
        const overtime = s.elapsed - s.duration
        if (this.tts === 'playing' ? overtime > Conversation.VOICE_GRACE : overtime >= 0) {
          this.setMode('idle')
        }
      }
    }
    this.level = approach(this.level, raw, 14, dt)
  }

  private trackTurn(raw: number, dt: number): void {
    if (raw > Conversation.SPEECH_LEVEL) {
      this.talkTime += dt
      this.silence = 0
    } else {
      this.silence += dt
    }
    if (!this.usingLiveMic) return
    this.liveTime += dt
    const finishedTalking = this.talkTime > 0.5 && this.silence > Conversation.END_OF_TURN_SILENCE
    const gaveUp = this.talkTime < 0.3 && this.liveTime > Conversation.NO_SPEECH_TIMEOUT
    if (finishedTalking || gaveUp) this.setMode('thinking')
  }

  private setMode(mode: Mode): void {
    if (mode === this.mode) return
    this.leave(this.mode)
    this.mode = mode
    this.modeTime = 0
    if (mode === 'listening') this.startListening()
    else if (mode === 'thinking') this.startThinking()
    else if (mode === 'speaking') this.startSpeaking()
    for (const cb of this.modeListeners) cb(mode)
  }

  private leave(mode: Mode): void {
    if (mode === 'listening') {
      this.mic?.stop()
      this.recognizer?.stop()
      this.usingLiveMic = false
    } else if (mode === 'speaking' && this.tts === 'playing') {
      this.speaker?.stop()
      this.tts = 'none'
    }
  }

  private startListening(): void {
    const session = ++this.session
    this.speech = null
    this.transcript = ''
    this.talkTime = this.silence = this.liveTime = 0
    this.sim = new MicSimulator(this.rng)
    const mic = this.mic
    if (mic) {
      mic
        .start()
        .then(() => {
          if (session === this.session && this.mode === 'listening') this.usingLiveMic = true
          else mic.stop()
        })
        .catch(() => {
          if (this.mic !== mic) return
          this.mic = null
          this.notice('Microphone blocked, simulating your voice')
        })
    }
    this.recognizer?.start((text) => {
      if (session === this.session) this.transcript = text
    })
  }

  private startThinking(): void {
    this.thinkLeft = uniform(this.rng, ...Conversation.THINK_SECONDS)
    const reply = chooseReply(this.rng, {
      name: this.name,
      talkTime: this.talkTime,
      transcript: this.transcript,
      last: this.lastTemplate,
    })
    this.lastTemplate = reply.template
    this.replyText = reply.text
  }

  private startSpeaking(): void {
    const session = ++this.session
    this.speech = new SpeechTimeline(this.replyText, this.rng, this.voice.rate)
    this.tts = 'none'
    if (!this.speaker) return
    this.tts = 'playing'
    const alive = () => session === this.session && this.mode === 'speaking'
    this.speaker.speak(this.replyText, this.voice, {
      onBoundary: (charIndex) => {
        if (alive()) this.speech?.syncToChar(charIndex)
      },
      onEnd: () => {
        if (!alive()) return
        // An "end" before anything could be heard means the voice never played.
        if ((this.speech?.elapsed ?? 0) < 0.3) this.voiceFailed()
        else this.tts = 'ended'
      },
      onError: () => {
        if (alive()) this.voiceFailed()
      },
    })
  }

  private voiceFailed(): void {
    this.tts = 'failed'
    this.speaker = null
    this.notice('Voice unavailable in this browser, showing captions only')
  }

  private notice(message: string): void {
    for (const cb of this.noticeListeners) cb(message)
  }
}
